File: first_pass/medium/Evaluate Reverse Polish Notation/evalRPN.js
const evalaute = function (a, b, operator) {
    switch (operator) {
        case "+": return a + b;
        case "-": return b - a;
        case "/": return Math.trunc(b / a);
        case "*": return a * b;
        default: throw new Error("Invalid Operator");
    }
}

/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function (tokens) {
    const stack = [];

    for (let i = 0; i < tokens.length; ++i) {
        if (!isNaN(tokens[i])) {
            stack.push(Number.parseInt(tokens[i]));
        } else {
            const operator = tokens[i];
            const a = stack.pop();
            const b = stack.pop();

            const c = evalaute(a, b, operator);

            stack.push(c);
        }
    }

    return stack.pop();
};

const testEvalRPNS = function (tokens, expected) {
    const result = evalRPN(tokens);
    if (result === expected) {
        console.log("Test Case Passed!")
    } else {
        console.log(`Test Case Failed: tokens: ${tokens}, result: ${result}, expected: ${expected}`)
    }
}

testEvalRPNS(["2", "1", "+", "3", "*"], 9)
testEvalRPNS(["4", "13", "5", "/", "+"], 6)
testEvalRPNS(["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"], 22)
testEvalRPNS(["4", "3", "-"], 1)

File: second_pass/4_stack/3_reverse_polish_notation/eval_rpn.py
from typing import List


class MySolution:
    operators = ["+", "-", "*", "/"]

    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        n = len(tokens)

        for i in range(n):
            if self.isNumber(tokens[i]):
                stack.append(int(tokens[i]))
            else:
                b = stack.pop()
                a = stack.pop()
                calc = self.calculate(a, b, tokens[i])
                stack.append(calc)

        return stack.pop()

    def isNumber(self, token: str) -> bool:
        return token not in self.operators

    def isOperator(self, token: str) -> bool:
        return token in self.operators

    def calculate(self, a: int, b: int, operator: str) -> int:
        if operator == "+":
            return a + b
        if operator == "-":
            return a - b
        if operator == "*":
            return a * b
        if operator == "/":
            return int(a / b)

    def testEvalRPN(self, tokens, expected):
        actual = self.evalRPN(tokens)
        if actual == expected:
            print("Test Case Passed!")
        else:
            print(f"Test Case Failed! tokens: {tokens}, actual: {actual}, expected: {expected}")


MySolution().testEvalRPN(["2", "1", "+", "3", "*"], 9)
MySolution().testEvalRPN(["4", "13", "5", "/", "+"], 6)
MySolution().testEvalRPN(["4", "13", "5", "/", "+"], 6)
MySolution().testEvalRPN(["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"], 22)
